Add README (#2)
ffiori/highlight-api-demos: README.md

# Highlight API 

The [Custom Highlight API](https://drafts.csswg.org/css-highlight-api-1/) extends the concept of [highlight pseudo-elements](https://drafts.csswg.org/css-pseudo-4/#highlight-pseudo-element) by providing a way for web developers to style the text of arbitrary range objects ([Ranges](https://dom.spec.whatwg.org/#interface-range) or [StaticRanges](https://dom.spec.whatwg.org/#interface-staticrange)), rather than being limited to the user agent defined `::selection`, `::inactive-selection`, `::spelling-error`, and `::grammar-error`. It provides a programmatic way of adding and removing highlights that do not affect the underlying DOM structure, but instead applies styles to text based on range objects, accessed via the `::highlight()` pseudo element. This is useful in a variety of scenarios, including editing frameworks that wish to implement their own selection, find-on-page over virtualized documents, multiple selection to represent online collaboration, or spellchecking frameworks.

This document intends to give a brief introduction to the Highlight API and how to use it, as well as compare performance and ease-of-use with a traditional way of highlighting text.

> ## How to use it

Reviewer: Oh, I see you have a section like that at the end of the document.

Could you move that up before the "How to Use it", so that readers don't get confused if they try to run the code samples as they read?

> ## How to use it

Author: Sounds good, I'll move it.

let highlight = new Highlight(range1);
highlight.add(range2);
```

In order to have any effect, custom highlights need to be registered into the highlight registry, represented by a [HighlightRegistry](https://drafts.csswg.org/css-highlight-api-1/#registration) object. The highlight registry is accessed via the `highlights` attribute of the CSS namespace, and represents all the custom highlights registered for the current global object’s associated Document. It is a map that relates custom highlight names to Highlight objects. The names are represented by Strings, and they are used to link a custom highlight to its [corresponding style](https://drafts.csswg.org/css-highlight-api-1/#styling-highlights) defined in the style section. For example:

```Javascript
<style>
  ::highlight(foo){
      color: blue;
  }
</style>
<script>
  let customHighlight = new Highlight();
  CSS.highlights.set('foo', customHighlight);
</script>
```

In the case above, a custom highlight is registered under the name 'foo', so all ranges contained in it will be styled with text color blue, as specified in the style section.

> Finally, the next example shows a simple way of highlighting selected text with the Highlight API:

Reviewer: ```suggestion
Finally, the next example shows a simple way of applying styles to the currently selected text with the Highlight API:
```

```Javascript
<style>
  ::highlight(example-highlight) {
    color: blue;
  }
</style>

<script>
  const highlightName = "example-highlight";

  function highlightSelectionWithAPI() {

>     let range = window.getSelection().getRangeAt(0);

Reviewer: ```suggestion
    let range = window.getSelection().getRangeAt(0).cloneRange();
```

We probably want to clone the range because per https://w3c.github.io/selection-api/#dom-selection-getrangeat the range returned by `getRangeAt()` is a reference, not a copy, of the one held inside the Selection implementation. It is a little suspicious to use that same range inside the Highlight, since a change that range would update both the user selection and the application of the ::highlight style.

In practice it basically works as-is because Selection seems to throw away its internal range whenever the user moves the selection. But I'd like to remove the ambiguity from the demo.

Author: Oh yes, I agree, good catch there, thanks!

    if(CSS.highlights.has("example-highlight")) 
      CSS.highlights.get(highlightName).add(range);
    else
      CSS.highlights.set(highlightName, new Highlight(range));
  }
  
  function clearAPIHighlight() {
    CSS.highlights.delete(highlightName);
  }
</script>
```

## How to highlight text *without* the Highlight API

> Traditionally, a common way to highlight different pieces of text dynamically in a webpage is to just wrap the content with a new `<span>` with a particular class or id that is given a specific style. An example of this technique can be found in [chromium code search](https://source.chromium.org)'s custom find on page, where if you look for text inside a source code file, the site highlights all matching strings with new spans. An actual screenshot is show next:

Reviewer: ```suggestion
Traditionally, a common way to highlight different pieces of text dynamically in a webpage is to just wrap the content with a new `<span>` with a particular class or id that is given a specific style. An example of this technique can be found in [chromium code search](https://source.chromium.org)'s custom find-on-page, where if you look for text inside a source code file, the site highlights all matching strings with new spans. An actual screenshot is shown next:

```

![Screenshot of real-life example of wrapping text with spans to highlight find-on-page results in Chromium Code Search](resources/custom-find-on-page-spans.png)

However, implementing this can be somewhat cumbersome because we need to manipulate the DOM tree by inserting new objects into it. This can lead to many issues:

- We should be careful not to break other processing of the DOM or cached elements, as the number of children/siblings of a node could change. This means the rest of the webpage should support dynamic changes in the DOM and act accordingly.
- Some cases are not easy to handle. For example, highlighting a chunk of text that crosses a node boundary.

> - Performance could be impacted due to the DOM growing too much. This could affect operations such as rendering frames or performing other actions needed by the site itself. See [Demo 2](https://github.com/ffiori/highlight-api-demos/tree/add-readme#demo-2-performance-of-both-approaches) for a performance comparison with the Highlight API.

Reviewer: ```suggestion
- Performance could be impacted due to the cost of allocating the `<span>`s and adding them to the DOM, including the cost of rerunning layout. This could affect operations such as rendering frames or performing other actions needed by the site itself. See [Demo 2](https://github.com/ffiori/highlight-api-demos/tree/add-readme#demo-2-performance-of-both-approaches) for a performance comparison with the Highlight API.

```

- Reverting the operation is not easy in non-trivial sites. We should traverse the DOM looking for spans with this specific id or class and delete them while reinserting their content into the same place it was before. Furthermore, this operation could be computationally expensive if the DOM's got too big and there are many highlighted spans to remove.

> 

Reviewer: Another thing we could mention is that the `<span>` method gets very complicated if we have multiple types of highlights we want to apply, especially if the different kinds can overlap. This is related to the point about highlighting chunks of text that cross node boundaries, since in this case we are *creating* node boundaries when we add spans.

Author: Thanks for mentioning this case, definitely useful 👍

Here is a possible way of implementing this technique to highlight selected text that would only work for simple sites:

```Javascript
<style>
  .highlighted-with-span {
    color: red;
  }
</style>

<script>

>   const highlightedWithSpanClass = "highlighted-with-span";

Reviewer: Same as previous example, for short code snippets consider just inlining these string literals.

  function highlightSelectionWithSpans() {
    if (window.getSelection().isCollapsed)
      return;
    try {
      let span = document.createElement('span');
      span.className = highlightedWithSpanClass;
      window.getSelection().getRangeAt(0).surroundContents(span);
    }

>     catch (error) {

Reviewer: How can this fail? Do we need the try/catch?

Reviewer: I see, could you just add that explanation as a short comment in the `catch` block?

Author: Sounds good, added a comment.

      ; // Treat error
    }
  }

  function clearSpans() {
    let spans = document.getElementsByClassName(highlightedWithSpanClass);
    while(spans.length) {
      let span = spans[0];
      span.parentNode.insertBefore(span.firstChild, span);
      span.parentNode.removeChild(span);
    }
  }
<script>
```

## Examples

> This section describes two examples of sites that make use of the Highlight API and intend to compare it to using spans to achieve the same goal. Note that, to be able to run the demos, Edge or Chrome >= 93 is needed (although it's recommended to use >= 95 for testing including the latest patches), and the browser should run with the flag `--enable-experimental-web-platform-features` enabled (it can be turned on in [edge://flags](edge://flags/#enable-experimental-web-platform-features) or [chrome://flags](chrome://flags/#enable-experimental-web-platform-features) under the name of 'Experimental Web Platform features').

Reviewer: ```suggestion
This section describes two demos that make use of the Highlight API and intend to compare it to using spans to achieve the same goal. Note that, to be able to run the demos, Edge or Chrome >= 93 is needed (although it's recommended to use >= 95 for testing including the latest patches), and the browser should run with the flag `--enable-experimental-web-platform-features` enabled (it can be turned on in [edge://flags](edge://flags/#enable-experimental-web-platform-features) or [chrome://flags](chrome://flags/#enable-experimental-web-platform-features) under the name of 'Experimental Web Platform features').

```

Reviewer: (Suggesting this change because I think the current wording implies that these are live sites "in the wild" that are already using the API, rather than something that we've just created).

Author: Makes sense, thanks!
